### PR TITLE
Make JavaFX a first-class citizen

### PR DESCRIPTION
**Pull request by tobiasdiez**
With this PR the main frame is converted to use JavaFX. Thus the role is now inverted and instead of embedding JavaFX in a Swing app, we now embedded the old Swing controls in a real JavaFX app. Hopefully, this leads to less problems.

The current state is very very basic and a lot of stuff is missing (side bar, menu, lila header instead of toolbar). The main reason for putting it out right now is to test if pushing the migration further fixes the issues encountered with Linux and MacOS.
@Siedlerchr @LinusDietz could you please try out this version and see if JabRef starts and shows a maintable (the latter only works if JabRef normally opens a bib file for you, since there is currently no GUI to open a new file).

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [ ] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Xbuntu 16.04
image

**Comment by tobiasdiez** [comment]
@Siedlerchr Thanks a lot for the active support! Only four more menus to go...

**Comment by Siedlerchr** [comment]
@tobiasdiez  What do we do with the actions that require a BasePanel instance? For Example the Content Selector Dialog

**Comment by tobiasdiez** [comment]
For these actions that work against the currently open database, I wrote a small wrapper `OldDatabaseCommandWrapper`. This is a bit ugly and should be changed in the long term but I didn't had the motivation to change all actions in `BasePanel`. So for the content selectors, you just pass `Actions.MANAGE_SELECTORS`.

**Comment by Siedlerchr** [comment]
I now converted the tools menu - only Push to external files and OpenOffice is missing. For the latter I was not sure how I should show that side pane (same applies to Web Search as well)

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 1]
```diff
@@ -1206,8 +1175,7 @@ private MenuBar createMenu() {
                 view,
                 tools,
                 options,
-                help
-        );
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/JabRefFrame.java]: @Siedlerchr Just a minor point: Is it possible to disable this particular auto-format of the braces in eclipse? I noticed similar reformattings also in other PRs of you. In my opinion, there are situation where it the code is easier to understand if the braces are on the next line (when it functions more as a grouping then as  a function argument)

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/JabRefFrame.java]: Hm., I will try to have a look at this.

**Comment by tobiasdiez** [comment]
This is the current state:
image

Everything you see is JavaFX and only two menus are missing. Concerning the color, we still have to experiment a bit (you are welcome to do so, just play around with `fx-theme-base` in `main.css`). Here in a typical business-blue:
image

**Comment by Siedlerchr** [comment]
Yeah! Which menus are missing? Then I could try to work on that

**Comment by tobiasdiez** [comment]
Library and Quality are the only ones missing.

**Comment by Siedlerchr** [comment]
I tried to convert some more actions, but still getting some NPE when I try to call frame.getCurrentBasePanel() inside the action...
That's the only showstopper atm.

File: JabRefModded2Profile.xml [hunk 2]
```diff
@@ -0,0 +1,315 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<profiles version="13">
+<profile kind="CodeFormatterProfile" name="JabRefModded2" version="13">
```
Comment (tobiasdiez) [inline JabRefModded2Profile.xml]: @Siedlerchr This should probably go in a separate PR, right?

**Comment by tobiasdiez** [comment]
This PR is now ready for review.
However, it is big enough to be completely unreviewable. Sorry for this, but converting the menu was not possible in smaller steps. I thus would propose to merge it (into the javafx branch) and let our users test it thoroughly. I clicked on every menu item to see if something happens and no exception is directly thrown. But of course is only a very shallow test and I'm pretty sure some functionality is broken with this PR.

Btw, nice side-effect: JabRefFrame has a 30% size decrease (-800 lines net, but still sits at 1700).

Thanks @Siedlerchr for the help! @person who merges: please don't squash the commits.

**Comment by Siedlerchr** [comment]
I will do play around with the menu items tomorrow and see if I encounter any erros.

**Comment by Siedlerchr** [comment]
I did an extensive test and fixed some issues I encountered.
The rest works as expected.
The only issue I have is staring:; The group panel pref? size is set to half of the window.
Resizing the window to the right makes the groups Panel to increase it's size, too. 
Oh, and when I decrease the window width from right to left, the groups panel seems to keep it's size

grafik
After resize: 
grafik


A more general thing which currently fails: Undo.
````
22:09:49.572 [JavaFX Application Thread] ERROR org.jabref.gui.BasePanel - runCommand error: field value must not be null
java.lang.NullPointerException: field value must not be null
	at java.util.Objects.requireNonNull(Objects.java:228) ~[?:1.8.0_162]
	at org.jabref.model.entry.BibEntry.setField(BibEntry.java:396) ~[main/:?]
	at org.jabref.model.entry.BibEntry.setField(BibEntry.java:442) ~[main/:?]
	at org.jabref.model.entry.BibEntry.setCiteKey(BibEntry.java:189) ~[main/:?]
	at org.jabref.gui.undo.UndoableKeyChange.undo(UndoableKeyChange.java:39) ~[main/:?]
	at javax.swing.undo.CompoundEdit.undo(CompoundEdit.java:63) ~[?:1.8.0_162]
	at javax.swing.undo.UndoManager.undoTo(UndoManager.java:335) ~[?:1.8.0_162]
	at javax.swing.undo.UndoManager.undo(UndoManager.java:413) ~[?:1.8.0_162]
	at org.jabref.gui.undo.CountingUndoManager.undo(CountingUndoManager.java:30) ~[main/:?]
	at org.jabref.gui.BasePanel$UndoAction.action(BasePanel.java:1969) ~[main/:?]
	at org.jabref.gui.BasePanel.runCommand(BasePanel.java:972) ~[main/:?]
	at org.jabref.gui.actions.OldDatabaseCommandWrapper.execute(OldDatabaseCommandWrapper.java:37) ~[main/:?]
	at org.jabref.gui.actions.JabRefAction.lambda$2(JabRefAction.java:23) ~[main/:?]
	at org.controlsfx.control.action.Action.handle(Action.java:419) ~[controlsfx-8.40.15-SNAPSHOT.jar:8.40.15.SNAPSHOT]
	at org.controlsfx.control.action.Action.handle(Action.java:64) ~[controlsfx-8.40.15-SNAPSHOT.jar:8.40.15.SNAPSHOT]
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191) ~[jfxrt.jar:?]
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49) ~[jfxrt.jar:?]
	at javafx.event.Event.fireEvent(Event.java:198) ~[jfxrt.jar:?]
	at javafx.scene.control.MenuItem.fire(MenuItem.java:462) ~[jfxrt.jar:?]
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1405) ~[jfxrt.jar:?]
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.lambda$createChildren$343(ContextMenuContent.java:1358) ~[jfxrt.jar:?]
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218) ~[jfxrt.jar:?]
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191) ~[jfxrt.jar:?]
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59) ~[jfxrt.jar:?]
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114) ~[jfxrt.jar:?]
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114) ~[jfxrt.jar:?]
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114) ~[jfxrt.jar:?]
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74) ~[jfxrt.jar:?]
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54) ~[jfxrt.jar:?]
	at javafx.event.Event.fireEvent(Event.java:198) ~[jfxrt.jar:?]
```

**Comment by lenhard** [comment]
For some reason, I am getting localization warnings when executing this branch, e.g.:
```
14:50:52.075 [JavaFX Application Thread] WARN  org.jabref.logic.l10n.Localization - Warning: could not get menu item translation for "Library" for locale en
```
Apart from that, a bigger thing that is still missing imho are most of the toolbar items.

We definitely should do one more release, before we merge this. Imho, this might even justify upgrading the version number to JabRef 5.0

**Comment by tobiasdiez** [comment]
@lenhard Which toolbar icons do you miss? I added all which I found reasonable to be placed that prominently (e.g., actions that the user does on a regular basis). Better have too few icons than too many, in my opinion.

**Comment by lenhard** [comment]
I tried to merge master into this and hope I did not break too many features on the way. We really need more frequent synchronization, otherwise there is no change of getting this together. Imho, this branch should be synced very urgently.

Here's the toolbars of this branch and master:
toolbar
That is a HUGE difference in available functionality. I think if we drop nearly all toolbar items, then many users will be pissed.

Added to that: The Web search sidebar needs to be rewritten as well. At the moment it basically crashes when you resize.

**Comment by koppor** [comment]
New toolbar:

grafik

Way too empty for me!!

Old toolbar:

grafik

I marked the things absolutely necesary for me. I don't use the other buttons. I toogle the interfaces via the full menu or via shortcut. copy and paste also not via these menus. Navigation also not via the toolbar.

As "good developers" we could have a user voting.

Or do we have Azure Analytics in place saying us which buttons are used?

Current opinion: Add my loved buttons and install Azure Analytics in parallel for that. Release 4.2 and collect statistics for JabRef 4.3.

**Comment by tobiasdiez** [comment]
@lenhard Thanks for merging with master, however, this PR is based on the other maintable migration branch. Now all those other changes in master occur as diff for this PR. Does everything work fine if we merge this PR now into the maintable branch?

Are the buttons for the shared databases needed? I always thought that changes are synced completely automatically. What extra value does the user gets from a facebook or github icon in the toolbar? How many users regularly open the terminal through JabRef? 
Probably we need an option to allow the users to decide if he wants some buttons or not.

In general, I'm hugely in favor of a data-based decision. 
We have some data about which dialogs are used how often. However, not every action is currently tracked (e.g. new library just opens a os-file dialog and is thus not captured in the list below; but also a few of our own dialogs are not tracked for some reason - for example the cleanup dialog).
The results below are nonetheless relatively clear and point in one direction: the majority of the users does not use any of our dialogs, except the "new entry" one and a few dialogs that are triggered automatically (import, duplicates, ...).

So I would propose to try-out a as minimal as possible version and then slowly add buttons as the users ask for them and/or our data shows that many users click on a certain action in the menu quite often.

name | count
-- | --
org.jabref.gui.EntryTypeDialog | 2640
org.jabref.gui.importer.ImportInspectionDialog | 1251
org.jabref.gui.DuplicateResolverDialog | 516
org.jabref.gui.mergeentries.MergeFetchedEntryDialog | 260
org.jabref.pdfimport.ImportDialog | 138
org.jabref.gui.preftabs.PreferencesDialog | 122
org.jabref.gui.groups.GroupDialog | 75
org.jabref.gui.collab.ChangeDisplayDialog | 40
org.jabref.gui.StringDialog | 26
org.jabref.gui.importer.FetcherPreviewDialog | 26
org.jabref.gui.mergeentries.MergeEntriesDialog | 24
org.jabref.gui.ReplaceStringDialog | 24
org.jabref.gui.help.NewVersionDialog | 22
org.jabref.gui.shared.ConnectToSharedDatabaseDialog | 20
org.jabref.gui.dbproperties.DatabasePropertiesDialog | 19
org.jabref.gui.customentrytypes.EntryCustomizationDialog | 16
org.jabref.gui.preftabs.FontSelectorDialog | 14
org.jabref.gui.GenFieldsCustomizer | 13
org.jabref.gui.MergeDialog | 12
org.jabref.gui.preftabs.PreferencesFilterDialog | 9
org.jabref.gui.externalfiles.SynchronizeFileField$OptionsDialog | 8
org.jabref.gui.auximport.FromAuxDialog | 8
org.jabref.gui.FindUnlinkedFilesDialog | 7
org.jabref.gui.plaintextimport.TextInputDialog | 7
org.jabref.gui.exporter.ExportCustomizationDialog | 6
org.jabref.gui.openoffice.StyleSelectDialog$AddFileDialog | 3
org.jabref.gui.externalfiletype.ExternalFileTypeEditor | 3
org.jabref.gui.externalfiles.WriteXMPAction$OptionsDialog | 2
org.jabref.gui.contentselector.ContentSelectorDialog | 2
org.jabref.gui.exporter.CustomExportDialog | 2
org.jabref.gui.importer.ImportCustomizationDialog | 1

**Comment by Siedlerchr** [comment]
We should discuss this in a devcall and maybe do a user survey?
From my point of view:
Cleanup, Generate BibtexKeys, Toogle Groups, Preview, WebSeach etc, push to application. 
Donate button. Help. Search field and Search mode changing
Save/Save All

**Comment by koppor** [comment]
You can easily merge it to the other branch. Git automatically recognizes existing commits of other branches.

- Social media: My aim is to get followers at these profiles. So I want to make it easy for users to follow us.
- Database: MySQL does not offer pub/sub, so you have to pull changes manually as outlined at http://help.jabref.org/en/SQLDatabase#usage
- Console: Wished by @simonharrer at https://github.com/JabRef/jabref/issues/462

I am unsure whether this is a chicken/egg thing. What happens if I start writing blog posts advertising our features? I think, most features are unused, because people just do not know about it.

Will be a good discussion at the devcall.

**Comment by tobiasdiez** [comment]
I merged this PR now to have a somewhat workable maintable-beta branch again, that is also head-to-head with the current master (thanks @lenhard !).

You all raised good points concerning the toolbar. I think this will be lively discussion in the next dev-call. I marked the related issue https://github.com/JabRef/jabref/issues/3678 with the corresponding label so that we don't forget it.
